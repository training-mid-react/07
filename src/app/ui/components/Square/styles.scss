.board__square {
    cursor: pointer;
    border: 1px solid #000;
    border-radius: 100%;

    &__content {
        width: 50px;
        height: 50px;
        border-radius: 100%;
    }
}

.falling {
    animation: fall 0.8s ease-out;
}

@keyframes fall {
    from {
        transform: translateY(-350px);
    }
    to {
        transform: translateY(0);
    }
}
