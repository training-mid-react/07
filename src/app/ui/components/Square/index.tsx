import { SquareProps } from '@core/types';
import React from 'react';
import './styles.scss';

export const Square: React.FC<SquareProps> = ({
    value,
    isFalling,
    onClick,
}) => {
    let isFallingClass = isFalling ? 'falling' : '';
    return (
        <div className="board__square">
            <div
                className={`board__square__content ${isFallingClass}`}
                style={{ backgroundColor: value ?? '' }}
                onClick={onClick}
            ></div>
        </div>
    );
};
