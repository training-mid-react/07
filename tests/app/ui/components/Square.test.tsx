import { fireEvent, render, screen } from '@testing-library/react';
import { Square } from '../../../../src/app/ui/components/Square/index';
import { vi } from 'vitest';
import React from 'react';
// agruegue esto y me dejo de dar error
import '@testing-library/jest-dom';

describe('Square', () => {
    const mockOnClick = vi.fn();
    const value: 'red' | 'blue' | null = null;
    const isFalling: boolean = false;

    test('Match snapshoot', () => {
        const { asFragment } = render(
            <Square onClick={mockOnClick} value={value} isFalling={isFalling} />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    test('Should render with value "red"', () => {
        const { container } = render(
            <Square onClick={mockOnClick} value="red" isFalling={isFalling} />
        );

        const square = container.firstChild;
        const squareContent = square?.firstChild;
        expect(square).toHaveClass('board__square');
        expect(squareContent).toBeInTheDocument();
        expect(squareContent).toHaveClass('board__square__content');
        expect(squareContent).toHaveStyle('background-color:  rgb(255, 0, 0)');
    });

    test('Should render with value "blue"', () => {
        const { container } = render(
            <Square onClick={mockOnClick} value="blue" isFalling={isFalling} />
        );

        const square = container.firstChild;
        const squareContent = square?.firstChild;
        expect(squareContent).toHaveStyle('background-color:  rgb(0, 0, 255)');
    });

    test('Should call onClick when clicked on square', () => {
        const { container } = render(
            <Square onClick={mockOnClick} value={value} isFalling={isFalling} />
        );

        const square = container.firstChild as Element;
        const squareContent = square?.firstChild as Element;

        fireEvent.click(squareContent);
        expect(mockOnClick).toHaveBeenCalled();
    });

    test('Should call onClick multiple times on multiple clicks', () => {
        const { container } = render(
            <Square onClick={mockOnClick} value={value} isFalling={isFalling} />
        );

        const squareContent = container.firstChild?.firstChild as HTMLElement;

        fireEvent.click(squareContent);
        fireEvent.click(squareContent);
        fireEvent.click(squareContent);

        expect(mockOnClick).toHaveBeenCalled();
    });

    test('renders the class falling when isFalling is true', () => {
        const { container } = render(
            <Square value="blue" onClick={mockOnClick} isFalling={!isFalling} />
        );

        const square = container.firstChild as Element;
        const squareContent = square?.firstChild as Element;

        fireEvent.click(squareContent);
        expect(mockOnClick).toHaveBeenCalled();
        expect(squareContent).toHaveClass('falling');
    });

    test('does not apply falling class when isFalling is false', () => {
        const { container } = render(
            <Square value="blue" isFalling={isFalling} onClick={mockOnClick} />
        );

        const square = container.firstChild as Element;
        const squareContent = square?.firstChild as Element;

        expect(squareContent).toHaveClass('board__square__content');
        expect(squareContent).not.toHaveClass('falling');
    });
});
