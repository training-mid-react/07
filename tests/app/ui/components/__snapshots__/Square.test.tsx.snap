// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Square > Match snapshoot 1`] = `
<DocumentFragment>
  <div
    class="board__square"
  >
    <div
      class="board__square__content "
    />
  </div>
</DocumentFragment>
`;
